source/server/hot_restart_impl.cc

```diff
@@ -26,11 +26,18 @@ namespace Server {
 // from working. Operations code can then cope with this and do a full restart.
 const uint64_t SharedMemory::VERSION = 9;
 
-SharedMemory& SharedMemory::initialize(Options& options) {
+static SharedMemoryHashSetOptions sharedMemHashOptions(Options& options) {
+  SharedMemoryHashSetOptions hash_set_options;
+  hash_set_options.capacity = options.maxStats();
+  hash_set_options.num_slots = Primes::findPrimeLargerThan(hash_set_options.capacity / 2);
```

Comment:
It's yak-shaving either way, most likely.  But my intuition is different.  Here's why.
- the penalty for the search on collisions is a strcmp, whereas the penalty for searching for a prime is a uint32 modulus.
- at (say) 10k maxKeys, I'm searching for a prime >5k and I only have to look at ~5 numbers 
- the current hash function may give a really good distribution without primes, but it's really sensitive and harder to reason about.

How about we leave this as is for now, and I'll put in a TODO to do a benchmark once https://github.com/envoyproxy/envoy/pull/2350 is in?

Reply:
I'm fine with that.  I wasn't as worried about the time spent finding a prime, as the added code complexity for probably minimal performance savings (especially since it is only the performance of loading config).  But let's wait and do a benchmark, then decide.

Reply:
How do primes come into this? Sorry not a hash table/set expert. If we keep the primes can you add some comments about how/whey they are used?

Reply:
TL;DR: better distribution for many hash functions.

It may not be needed for the hash function I found in Envoy, but let's say were making a hash table using pointers as keys.  The pointers would typically have the low order 3 bits 0, so you'd only use 1/8 of your buckets.  modding with a prime solves that and gets you a better distribution.

https://stackoverflow.com/questions/3980117/hash-table-why-size-should-be-prime

I just left that as a one-line comment at the call-site.

Reply:
+1, thanks